### PR TITLE
Process events received through Nuts Network

### LINKED ISSUE #141: Process events received through Nuts Network
Events received through Nuts Network should be processed as if they would've been received through GitHub.

This however requires changes to the way events are ordered and processed because currently the event system assumes events are received in order. This is not necessarily the case for events received through Nuts Network. As such, we should keep track of which events we received (by hash) and keep an ordered list of events which we couldn't apply yet because we were missing preceding events.

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes #141 

This PR is two-fold:

1. Support processing registry events received out of order but whose timestamps are chronologically in order. Nuts Network delivers documents in unordered, which can cause problems when processing registry events (e.g. endpoint registration before organization exists). This PR allows failed events to be retried later using a chronologically ordered retry 'queue'.
2. Subscribe to network documents of type **nuts.registry-event** (which indicates a registry event) and process them.

### CODE REVIEW

**Comment by woutslakhorst** [comment]
Is it possible to get an event that never succeeds and will be stuck in retry?

**Comment by woutslakhorst** [comment]
And am I correct to assume that an event is retried when a relating event is received?

File: api/client_test.go [hunk 1]
```diff
@@ -73,6 +73,13 @@ func TestHttpClient_OrganizationById(t *testing.T) {
 			t.Errorf("Expected return organization identifier to be [%s], got [%s]", organizations[0].Identifier, res.Identifier)
 		}
 	})
+
+	t.Run("http execution error", func(t *testing.T) {
```
Comment (woutslakhorst) [inline api/client_test.go:77]: Is it needed to do this test 6 times?

File: pkg/network/ambassador_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/nuts-foundation/nuts-registry/test"
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"os"
+	"sync"
+	"testing"
+	"time"
+)
+
+func Test_ambassador_SendAndReceive(t *testing.T) {
```
Comment (woutslakhorst) [inline pkg/network/ambassador_test.go:18]: This is a fairly large test, can it be split up?

File: pkg/registry_functional_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
+// +build !race
+
+/*
+ * Nuts registry
+ * Copyright (C) 2019. Nuts community
```
Comment (woutslakhorst) [inline pkg/registry_functional_test.go:5]: 2020

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
Could you add metrics for the size of the retry queue?

**Comment by reinkrul** [comment]
> Is it possible to get an event that never succeeds and will be stuck in retry?

Yes that's possible. However since queues are in-memory Go channels, they'll be gone after a restart. For now that's not a problem since data will (should) be published on Github X time after it was published on the network.

**Comment by reinkrul** [comment]
> And am I correct to assume that an event is retried when a relating event is received?

Yes and no. Yes: after an event was processed, any pending events are retried. As such when the related event (which preceeds the event to be retried in this case) is received and processed, the event set aside earlier (because it was received out of order) is retried. No: it's not a "smart" algorithm that it looks up which events it should retry (this could/should be added later), it just orders events chronologically (using the `createdAt` timestamp established by the creator of the event) because that's how they're ordered when published on Github.

For consistent decentralized ordering, the ref/prev event referencing should be used imo.

File: api/client_test.go (hunk 1 above)
Comment (reinkrul) [reply api/client_test.go:77]: Code (call HTTP client, check err) is there 6 times (actually more) and we needed more project test coverage, so yes... :-(

**Review by woutslakhorst — APPROVED** [review]

**Pull request merged**